### ISSUE TITLE
SIGQUIT crash

### ISSUE THREAD

**Issue by mzealey** [haproxy/haproxy#1782]
### Detailed Description of the Problem

I reloaded a running haproxy twice, and then SIGQUIT'd the oldest of the children to free up memory. Went from ~5gb to ~800mb and then crashed a few seconds later

```
Jul 12 16:36:52 haproxy-05 haproxy: Dumping pools usage. Use SIGQUIT to flush them.
Jul 12 16:36:52 haproxy-05 haproxy: - Pool cache_st (32 bytes) : 100912 allocated (3229184 bytes), 100912 used (~5375 by thread caches), needed_avg 101935, 0 failures, 6 users, @0xa18080 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool resolv_requ (64 bytes) : 398338 allocated (25493632 bytes), 398338 used (~5378 by thread caches), needed_avg 399361, 0 failures, 11 users, @0xa14e00 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool vars (96 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 4 users, @0xa1a200 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool ssl_sock_ct (128 bytes) : 244210 allocated (31258880 bytes), 244210 used (~5370 by thread caches), needed_avg 245233, 0 failures, 3 users, @0xa0d840 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool task (160 bytes) : 78063 allocated (12490080 bytes), 78063 used (~8650 by thread caches), needed_avg 79086, 0 failures, 2 users, @0xa1b2c0 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool h2s (192 bytes) : 100910 allocated (19374720 bytes), 100910 used (~5364 by thread caches), needed_avg 101933, 0 failures, 6 users, @0xa0e900 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool tcpcheck_ru (224 bytes) : 53132 allocated (11901568 bytes), 53132 used (~5364 by thread caches), needed_avg 54155, 0 failures, 4 users, @0xa11b80 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool ssl_sock_cl (256 bytes) : 16 allocated (4096 bytes), 16 used (~16 by thread caches), needed_avg 12, 0 failures, 3 users, @0xa0c780 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool spoe_ctx (320 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0xa13d40 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool resolv_reso (480 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0xa15ec0 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool resolv_answ (608 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0xa16f80 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool requri (1024 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0xa32280 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool stream (1152 bytes) : 53139 allocated (61216128 bytes), 53139 used (~5371 by thread caches), needed_avg 54162, 0 failures, 1 users, @0xa12c40 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool fcgi_conn (1216 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0xa10ac0 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool h2c (1312 bytes) : 12 allocated (15744 bytes), 12 used (~12 by thread caches), needed_avg 8, 0 failures, 1 users, @0xa0fa00 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool hpack_tbl (4096 bytes) : 12 allocated (49152 bytes), 12 used (~12 by thread caches), needed_avg 8, 0 failures, 1 users, @0x42b7e80 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool buffer (16384 bytes) : 76 allocated (1245184 bytes), 76 used (~76 by thread caches), needed_avg 72, 0 failures, 1 users, @0x42aed80 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: - Pool trash (16416 bytes) : 16 allocated (262656 bytes), 16 used (~16 by thread caches), needed_avg 12, 0 failures, 1 users, @0xa21540
Jul 12 16:36:52 haproxy-05 haproxy: - Pool dns_msg_buf (65568 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0xa19140 [SHARED]
Jul 12 16:36:52 haproxy-05 haproxy: Total: 19 pools, 166541024 bytes allocated, 166541024 used (~12583200 by thread caches).
Jul 12 16:36:55 haproxy-05 haproxy: Thread 1 is about to kill the process.
Jul 12 16:36:55 haproxy-05 haproxy: *>Thread 1 : id=0x7fb7bb27f9c0 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/1    stuck=1 prof=0 harmless=0 wantrdv=1
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=38901850213545 now=38903983745310 diff=2133531765
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: call trace(15):
Jul 12 16:36:55 haproxy-05 haproxy: | 0x7fb7ba6108e0 [48 c7 c0 0f 00 00 00 0f]: libpthread:+0x118e0
Jul 12 16:36:55 haproxy-05 haproxy: | 0x7fb7b95e34f7 [48 3d 01 f0 ff ff 73 01]: libc:madvise+0x7/0x21
Jul 12 16:36:55 haproxy-05 haproxy: | 0x7fb7b957a76a [ba 01 00 00 00 4c 8b 4c]: libc:malloc_trim+0x20a/0x28a
Jul 12 16:36:55 haproxy-05 haproxy: |       0x58d383 [48 83 c4 08 e9 74 dc ff]: pool_gc+0x53/0x63
Jul 12 16:36:55 haproxy-05 haproxy: |       0x59b4db [48 39 eb 48 8b 03 48 89]: __signal_process_queue+0x8b/0x128
Jul 12 16:36:55 haproxy-05 haproxy: Thread 2 : id=0x7fb7b6cb1700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/2    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39740610186700 now=39742932583185 diff=2322396485
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 3 : id=0x7fb7b62b0700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/3    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39560956487989 now=39563076630162 diff=2120142173
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 4 : id=0x7fb7825b8700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/4    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39192298900584 now=39194483882142 diff=2184981558
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 5 : id=0x7fb781bb7700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/5    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=40188491842848 now=40190784009395 diff=2292166547
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 6 : id=0x7fb7811b6700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/6    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39114062043645 now=39116215727660 diff=2153684015
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 7 : id=0x7fb7807b5700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/7    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39361206091437 now=39363505872493 diff=2299781056
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 8 : id=0x7fb77fdb4700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/8    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=38806118132808 now=38807885973161 diff=1767840353
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 9 : id=0x7fb77f3b3700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/9    stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39903363461731 now=39905548267141 diff=2184805410
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 10: id=0x7fb77e9b2700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/10   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=43130038923564 now=43132401898349 diff=2362974785
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 11: id=0x7fb77dfb1700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/11   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39041421697875 now=39043452503439 diff=2030805564
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 12: id=0x7fb77d5b0700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/12   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=39444869131362 now=39446925227691 diff=2056096329
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 13: id=0x7fb77cbaf700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/13   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=38633774639973 now=38635783889741 diff=2009249768
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 14: id=0x7fb77c1ae700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/14   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=38729564674733 now=38731536429229 diff=1971754496
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 15: id=0x7fb77b7ad700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/15   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=48530126447276 now=48532585734423 diff=2459287147
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:55 haproxy-05 haproxy: Thread 16: id=0x7fb77adac700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
Jul 12 16:36:55 haproxy-05 haproxy: 1/16   stuck=0 prof=0 harmless=1 wantrdv=0
Jul 12 16:36:55 haproxy-05 haproxy: cpu_ns: poll=40915522867992 now=40917333885180 diff=1811017188
Jul 12 16:36:55 haproxy-05 haproxy: curr_task=0
Jul 12 16:36:56 haproxy-05 haproxy: [NOTICE]   (10639) : haproxy version is 2.5.7-2ef551d
Jul 12 16:36:56 haproxy-05 haproxy: [NOTICE]   (10639) : path to executable is /usr/local/sbin/haproxy
Jul 12 16:36:56 haproxy-05 haproxy: [WARNING]  (10639) : Former worker (3363) exited with code 134 (Aborted)
```

### Expected Behavior

It should free memory but not crash. Perhaps 50% of the time I try this it crashes, usually on the oldest, biggest processes, which are precisely the ones I want to periodically free memory from.

### Steps to Reproduce the Behavior

It's not 100% reproducable, but SIGQUIT a long, heavily-running haproxy process and it often happens

### What is your configuration?

```haproxy
Unsure that it's relevant but happy to share if needed
```

### Output of `haproxy -vv`

```plain
$ haproxy -vv
HAProxy version 2.5.7-2ef551d 2022/05/13 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.7.html
Running on: Linux 5.10.118-111.515.amzn2.x86_64 #1 SMP Wed May 25 22:12:19 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_LIBCRYPT=1 USE_CRYPT_H=1 USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.3.1 20180712 (Red Hat 7.3.1-15)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```
```


### Last Outputs and Backtraces

_No response_

### Additional Information

_No response_

**Comment by wtarreau** [comment]
malloc_trim() took far too much time to complete. That may happen on some implementations and was already reported once. We've added a specific option for this, `no-memory-trimming` to place in the global section. It will prevent malloc_trim() from being called (and will release less memory as a side effect).

**Comment by mzealey** [comment]
Thanks, is there an allocator you would recommend running haproxy with - for example tcmalloc? And/or is it possible to raise the watchdog timer for this - as it's an old process we don't care too much about a few seconds of lockup, more about releasing as much memory as possible.

**Comment by wtarreau** [comment]
The best results we've got so far are with jemalloc. That's even what we're using in our ALOHA appliances at work. It's both significantly faster and less memory hungry than the default one from glibc.

I agree with your point regarding only a few seconds of lockup. The problem is that there's no way to know precisely what's causing the lockup, and we've had plenty of issues at stopping time that were resolved thanks to the watchdog, so I wouldn't want to disable it on stopping. Also we've seen malloc_trim() being killed by the user after several minutes in the past, due to strong competition against other threads that were allocating. Thus it seems it's not always that good, unfortunately. Ideally we'd like to pass it a small ratio of memory to release (i.e. no more than 1% per call). But that's not how it works, instead it takes the amount you want to keep, which is very vague.

With jemalloc this has never been a problem for us, it makes extensive use of mmap() and punches holes in the memory space whenever possible so that you don't end up with gigs of unused memory still allocated.
